client/src/lib.js

```diff
@@ -112,6 +116,71 @@ class NFTStorage {
     }
   }
 
+  /**
+   * @template {API.TokenInput} T
+   * @param {API.Service} service
+   * @param {T} data
+   * @returns {Promise<API.StoreResult<T>>}
+   */
+  static async store(
+    { endpoint, token },
+    { name, description, image, properties, decimals, localization }
+  ) {
+    const url = new URL(`/api/store`, endpoint)
+    // Just validate that expected field are present
+    if (typeof name !== 'string') {
+      throw new TypeError(
+        'string property `name` identifying the asset is required'
+      )
+    }
+    if (typeof description != 'string') {
+      throw new TypeError(
+        'string property `description` describing asset is required'
+      )
+    }
+
+    if (!(image instanceof Blob) || !image.type.startsWith('image/')) {
+      throw new TypeError(
+        'proprety `image` must be a Blob or File object with `image/*` mime type'
+      )
+    }
+    if (typeof decimals !== 'undefined' && typeof decimals != 'number') {
+      throw new TypeError('proprety `decimals` must be an integer value')
+    }
+
+    const body = new FormData()
+    const data = Token.encode(
+      { name, description, image, properties, decimals, localization },
+      body
+    )
+
+    body.set('meta', JSON.stringify(data))
+
+    const response = await fetch(url.toString(), {
+      method: 'POST',
+      headers: NFTStorage.auth(token),
+      body,
+    })
+
+    /** @type {API.StoreResponse<T>} */
+    const result = await response.json()
+
+    if (result.ok === true) {
+      const { value } = result
+      const data = Token.decode(value.data)
+
+      return {
+        ipld: CID.parse(value.ipld),
+        metadata: new URL(value.metadata.href),
+        data,
+        embed: Token.embed(data, {
+          gateway: GATEWAY,
+        }),
+      }
```

Comment:
What about this return type

```ts
interface Token {
  cid: CID
  medataURL: URL
  toJSON(options?: { gateway?: URL }):TokenJSON
}
```

Reply:
Context: @hugomrdias did not liked that data and embed were basically same data but with different URL encodings. We have talked bunch on what a better representation would be and I've wrote it down here for further discussion.